R/user_error_checks.R

```diff
@@ -0,0 +1,45 @@
+# This function exists to check if a portfolio has a single holdings
+# date, or multiple. If there is a single date, then everything is good.
+# but on the platform, if a user groups multiple  portfolios, and some of
+# thse have different holding dates, then the parameters file will
+# contaain a yaml array with the distinct dates from each grouped
+# portfolio.
+# Examples:
+#
+# simple portfolio
+# holdings_date: 2020Q4
+#
+# groupedd portfolio with all same date:
+# holdings_date: 2020Q4
+#
+# grouped portfolio with different dates:
+# holdings_date: [2019Q4, 2020Q4]
+#
+# `config` will read this in as a character vector, so all we need to do
+# here is check for length > 1
+
+check_grouped_portfolio_years <- function(
+  port_holdings_date
+  ) {
+  if ((!is.null(port_holdings_date)) && length(port_holdings_date) > 1) {
+    error_name <- "Grouped Portfolio with different Holdings Dates"
+    desc <- paste(
+      "When grouping portfolios, the user included portfolios with",
+      "different holdings dates. All portfolios in the group should have",
+      "the same holding date (2019Q4, 2020Q4, etc.) The portfolios",
+      "in this group have holding dates of:",
+      paste0("**", paste(port_holdings_date, collapse = ", "), "**")
+    )
+    action <- paste(
+      "Delete this grouping, and create a new one with portfolios",
+      "with the same holdings date before running the analysis again."
+    )
+    log_user_errors(
+      error_name = error_name,
+      suggested_action = action,
+      description = desc,
+      immediate = TRUE
+    )
+    stop("port_holdings_date length > 1")
```

Comment:
Nit: Maybe move this to the top so it's one case less we need to think about. Then we can focus on the longer happy path.

Nit: Maybe use `call. = FALSE` as that info clutters the console and can be retrieved easily with `traceback()`.

> Use stop(call. = FALSE), rlang::abort(), Rf_errorcall(R_NilValue, ...) to avoid cluttering the error message with the name of the function that generated it. That information is often not informative, and can easily be accessed via traceback() or an IDE equivalent.

https://style.tidyverse.org/error-messages.html

Reply:
The "unhappy" path (where there is an error) is the long one.

I generally prefer structures where it's along the lines of 
```r
x <- function() {
  if (condition) {
    stop("condition")
  }
  return(value)
}
```

over constructions like

```r
x <- function() {
  if (!condition) {
    return(value)
  }
  stop("condition")
  # it feels weird to have the last line of the function not be return()
}
```

There are cases where the second construction is better, but in general, if I'm looking for what the function gives me in the happy path (no `stop()`s or similar errors), the ffirst place I'm going to look is the last line of the function.

Reply:
Oh, I see -- this `check_*()` function writes logs before it aborts. Then it makes sense what you say. I guess I wouldn't have missed if the  details were extracted away the the two main actions are clearer: Log and abort:

```r
f <- function(first_argument) {
  if (something_happens) {
    log()
    abort()
  }

  invisible(first_argument)
}
```